### PR TITLE
Add jax.scipy.linalg.invpascal, invhilbert, fiedler_companion

### LINKED ISSUE #10144: Pascal, Hilbert and/or other special matrices in `jax.scipy.linalg`
Hi! 

`scipy.linalg` implements special matrices. At the moment, `jax.scipy.linalg` does not do that, but it would be useful for me to have access to Pascal and Hilbert matrices via `jax`. (I need them for some state-space modelling stuff, if that is relevant information.)

I could make a pull request if it helps :)
As a starting point, I will explain what I am doing currently below, and I will also explain some difficulties with `jit`ting that might need to be resolved before a pull request? 

## Related issues 

* https://github.com/google/jax/issues/1646 (vaguely related I guess)
* https://github.com/google/jax/issues/3171 (see below for why this is related)

## Hilbert matrix

For example, a Hilbert matrix would be rather straightforward to implement, via 
```python
def hilbert(n):
    a = jnp.arange(n)
    return 1 / (a[:, None] + a[None, :] + 1)
```
A problem I could see with this implementation is appropriate `jit` and `vmap` behaviour: the size of the output depends on `n` :( . This is true for most special matrices, unfortunately.

Therefore, in my own code, I have split the function into something like
```python

def hilbert(n):
    a = jnp.arange(n)
    return _hilbert(a)

@jax.jit
def _hilbert(a):
    return 1 / (a[:, None] + a[None, :] + 1)
```
which at least allows to `jit` parts of the implementation. 
An alternative would be to provide a Hankel matrix (where the shape of the matrix is determined by the shapes of the input vectors, and which should be easier to `jit`), and then let users like me do the Hilbert-via-Hankel conversion depending on each application. Scipy's Hankel matrix, however, uses `np.lib.as_strided`, and I understand from issue #3171 that this is not intended to be supported.
But this can probably be resolved, if one decides to go down that road.

## Pascal matrix

Pascal matrices are a little less obvious to implement than Hankel/Hilbert matrices, because they use binomial coefficients. For my applications, I have so far computed those via `jax.lax.exp(jax.lax.lgamma(n+1.))` (because `jnp.prod(jnp.arange(1, n+1))` has `jit` issues again; if I remember correctly, I took the gamma-function-idea from `jet.py`) and then vmapped a factorial implementation via 
```python
def pascal(n):
    return _pascal(jnp.arange(n))

@jax.jit
def _pascal(a):
    return _binom(a[:, None], a[None, :])


def _broadcast_to_matrix(k):
    """Take a function f(scalar, scalar)-> scalar and return a function g((n,m) array, (m, k) array) -> (n,k) array"""
    k_vmapped_x = jax.vmap(k, in_axes=(0, None), out_axes=-1)
    k_vmapped_xy = jax.vmap(k_vmapped_x, in_axes=(None, 1), out_axes=-1)
    return jax.jit(k_vmapped_xy)


@_broadcast_to_matrix
@jax.jit
def _binom(n, k):
    a = _factorial(n)
    b = _factorial(n - k)
    c = _factorial(k)
    return a / (b * c)


@jax.jit
def _factorial(n):
    return jax.lax.exp(jax.lax.lgamma(n + 1.0))
```
This is slightly different to `scipy.linalg.pascal`, however, because my implementation returns only floats and a general Pascal matrix should do integers, at least for `scipy.linalg.pascal(..., exact=True)`. And the implementation would be fairly different to how scipy does it, which might also not be desired.

## Long story short
Hilbert and Pascal matrices would be useful to have, and if the jitting issues can be resolved (or do not matter too much?) I can make a pull request. 

What do you think?

### PR DESCRIPTION
**Pull request by adityasingh2400**
Closes #10144.

Adds three more special-matrix constructors to `jax.scipy.linalg`, continuing the series from the recently merged `companion`, `fiedler`, and `pascal`/`hilbert` work.

### Functions

- `invpascal(n, kind)` returns the inverse of the Pascal matrix in each of the three kinds. The lower-kind closed form is `(-1)^(i - j) * binom(i, j)`, the upper kind is its transpose, and the symmetric inverse is `invp_U @ invp_L`. This matches scipy elementwise for `kind in {'symmetric', 'lower', 'upper'}` under floating point comparison.
- `invhilbert(n)` uses the standard binomial-coefficient closed form `(-1)^(i + j) * (i + j + 1) * comb(n + i, n - j - 1) * comb(n + j, n - i - 1) * comb(i + j, i)^2`. This matches scipy under floating point.
- `fiedler_companion(a)` builds the pentadiagonal Fiedler companion of a polynomial. The body is wrapped in `jnp.vectorize` so `(..., N)` inputs yield `(..., N - 1, N - 1)` outputs.

### scipy parity notes

- Neither `invpascal` nor `invhilbert` supports `exact=True`. JAX does not support arbitrary-precision integers, so both return float arrays. The docstrings spell this out and point users at scipy when they need exact integer entries. Parity tests therefore compare against scipy with `exact=False`.
- `fiedler_companion`, like the just-merged `companion`, does not raise at runtime when `a[..., 0] == 0`. Doing so would force concretization under jit, so the function lets the `inf` / `nan` from the leading-coefficient division propagate.

### Tests

- `testInvpascal`: parity vs scipy across `n in {0, 1, 2, 5, 8}` and all three kinds (15 cases, x32 and x64).
- `testInvhilbert`: parity vs scipy across `n in {0, 1, 2, 3, 5, 8}` (6 cases, x32 and x64).
- `testFiedlerCompanion`: parity vs scipy across 1-D and batched shapes for float dtypes, plus the existing `assertRaisesRegex` pattern for `N < 2`.
- Doctests for the three new functions pass.
- Full `tests/linalg_test.py` run is green (912 passed, 23 unrelated skips).

### Files

- `jax/_src/scipy/linalg.py`: three new functions plus a small `_fiedler_companion` helper.
- `jax/scipy/linalg.py`: re-exports.
- `docs/jax.scipy.rst`: adds the three symbols to the autosummary.
- `CHANGELOG.md`: Unreleased entries.
- `tests/linalg_test.py`: new parametrized tests and a small `osp_linalg_fiedler_companion` helper to mirror the pattern used for `companion` / `fiedler` / `leslie`.

### CODE REVIEW

File: jax/_src/scipy/linalg.py [hunk 1]
```diff
@@ -2730,6 +2730,84 @@ def fiedler(a: ArrayLike) -> Array:
...
+    >>> a = jnp.array([1., -16., 86., -176., 105.])
+    >>> jax.scipy.linalg.fiedler_companion(a)
+    Array([[ 16., -86.,   1.,   0.],
+           [  1.,   0.,   0.,   0.],
+           [  0., 176.,   0., -105.],
+           [  0.,   1.,   0.,   0.]], dtype=float32)
+  """
+  a, = promote_args_inexact("fiedler_companion", a)
+  a = jnp.atleast_1d(a)
+  if a.shape[-1] < 2:
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py]: We should accept a length-1 array here and return a `(0, 0)` matrix.

Comment (jakevdp) [reply jax/_src/scipy/linalg.py]: Update, I sent a PR to SciPy to make the length-1 case consistent with the function's documentation: https://github.com/scipy/scipy/pull/25227

File: jax/_src/scipy/linalg.py [hunk 2]
```diff
@@ -2730,6 +2730,84 @@ def fiedler(a: ArrayLike) -> Array:
...
+    raise ValueError(
+        "The length of `a` along the last axis must be at least 2; "
+        f"got shape {a.shape}.")
+  return _fiedler_companion(a)
+
+@partial(jnp_vectorize.vectorize, signature="(n)->(m,m)")
+def _fiedler_companion(a: Array) -> Array:
+  a = a / a[0]
+  n = a.shape[0] - 1
+  if n == 1:
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py:2782]: We should explicitly handle `n == 0` here, returning `jnp.empty_like(a, shape=(0, 0))`

File: jax/_src/scipy/linalg.py [hunk 3]
```diff
@@ -2730,6 +2730,84 @@ def fiedler(a: ArrayLike) -> Array:
...
+  j = jnp.arange(n)[None, :]
+  # Clip indices so out-of-range lookups are safe; masks below select valid
+  # cells.
+  super_val = -a[jnp.minimum(i + 2, n)]
+  sub_val = -a[jnp.minimum(i + 1, n)]
+  i_even = (i % 2 == 0)
+  i_odd = (i % 2 == 1)
+  one = jnp.array(1, dtype=a.dtype)
+  zero = jnp.array(0, dtype=a.dtype)
+  result = jnp.where(
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py]: This is an inefficient way to do multiple comparisons; please use `jnp.select` instead, which lowers to `lax.select_n`.

File: jax/_src/scipy/linalg.py [hunk 4]
```diff
@@ -2831,6 +2909,61 @@ def hilbert(n: int) -> Array:
...
+    >>> with jnp.printoptions(precision=3, suppress=True):
+    ...   print(jax.scipy.linalg.invhilbert(2))
+    ...   print(jax.scipy.linalg.invhilbert(3))
+    [[ 4. -6.]
+     [-6. 12.]]
+    [[   9.  -36.   30.]
+     [ -36.  192. -180.]
+     [  30. -180.  180.]]
+  """
+  from jax._src.scipy.special import comb
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py]: Please put imports at the top of the file.

File: tests/linalg_test.py [hunk 5]
```diff
@@ -2745,6 +2774,20 @@ def testHilbert(self, n):
     self._CheckAgainstNumpy(osp_fun, jsp_fun, args_maker)
     self._CompileAndCheck(jsp_fun, args_maker)
 
+  @jtu.sample_product(
+    n=[0, 1, 2, 3, 5, 8],
+    )
+  @jax.default_matmul_precision("float32")
```
Comment (jakevdp) [inline tests/linalg_test.py]: I don't think this is relevant here, because the implementation doesn't use any matmuls.

**Review by jakevdp — CHANGES_REQUESTED** [review]

**Comment by jakevdp** [comment]
Also, please follow the advice of the gemini comments regarding default float dtype in two places. Thanks!

**Comment by adityasingh2400** [comment]
Done in a783b3d7a. invhilbert and invpascal now build their index array with dtypes.default_float_dtype() instead of hardcoded np.float32, so both honor the x64 config like the rest of the module.

**Comment by jakevdp** [comment]
See also the inline comments above – thanks!

**Comment by adityasingh2400** [comment]
Thanks, addressed all the inline points in 4433748d9:

- fiedler_companion now returns an empty (0, 0) matrix for fewer than two coefficients instead of raising, and _fiedler_companion handles n == 0 with jnp.empty_like. Thanks for filing scipy#25227, the behavior here matches what you proposed there.
- Replaced the nested jnp.where with jnp.select so it lowers to lax.select_n.
- Moved the scipy.special.comb import to the top of the file.
- Dropped the default_matmul_precision decorator from testInvhilbert since invhilbert has no matmul. I kept it on testInvpascal because the symmetric case computes jnp.dot(L_inv.T, L_inv).

File: jax/_src/scipy/linalg.py [hunk 6]
```diff
@@ -2730,6 +2731,79 @@ def fiedler(a: ArrayLike) -> Array:
...
+  a, = promote_args_inexact("fiedler_companion", a)
+  a = jnp.atleast_1d(a)
+  return _fiedler_companion(a)
+
+@partial(jnp_vectorize.vectorize, signature="(n)->(m,m)")
+def _fiedler_companion(a: Array) -> Array:
+  n = a.shape[0] - 1
+  if n == 0:
+    return jnp.empty_like(a, shape=(0, 0))
+  a = a / a[0]
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py:2781]: Somewhere we should check if `a.shape[0] == 0` and raise a `ValueError`, otherwise we'll get an `IndexError` in the implementation here.

**Review by jakevdp — COMMENTED** [review]
One last comment – after you address it, could you please squash all changes into a single commit? Thanks!

**Comment by adityasingh2400** [comment]
Thanks for the careful review. I added a guard in the public `fiedler_companion` right after `atleast_1d`: if the last axis has length 0 it now raises a clear `ValueError` (with the shape in the message) instead of hitting the `IndexError` from `a / a[0]`. The length-1 case still returns a `(0, 0)` matrix as before, and only length-0 raises. I also added a quick test for the empty-axis case. And I squashed everything down to a single commit. Let me know if you'd like any other changes.

**Review by jakevdp — APPROVED** [review]
Thanks!
